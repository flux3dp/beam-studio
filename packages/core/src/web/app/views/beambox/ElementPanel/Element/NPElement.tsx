import React, { useContext, useState } from 'react';

import classNames from 'classnames';

import progressCaller from '@core/app/actions/progress-caller';
import { ElementPanelContext } from '@core/app/contexts/ElementPanelContext';
import HistoryCommandFactory from '@core/app/svgedit/history/HistoryCommandFactory';
import undoManager from '@core/app/svgedit/history/undoManager';
import importSvg from '@core/app/svgedit/operations/import/importSvg';
import postImportElement from '@core/app/svgedit/operations/import/postImportElement';
import { getNPIconByID } from '@core/helpers/api/flux-id';
import { getData } from '@core/helpers/layer/layer-config-helper';
import { getLayerByName } from '@core/helpers/layer/layer-helper';
import { getSVGAsync } from '@core/helpers/svg-editor-helper';
import webNeedConnectionWrapper from '@core/helpers/web-need-connection-helper';
import type { IIcon } from '@core/interfaces/INounProject';
import type ISVGCanvas from '@core/interfaces/ISVGCanvas';

import styles from './Element.module.scss';

const progressId = 'import-noun-project-svg';
let svgCanvas: ISVGCanvas;

getSVGAsync((globalSVG) => {
  svgCanvas = globalSVG.Canvas;
});

const importNPSvg = async (id: string) => {
  progressCaller.openNonstopProgress({ id: progressId });

  try {
    const base64 = await getNPIconByID(id);

    if (!base64) return;

    const res = await fetch(base64);
    const blob = await res.blob();

    const batchCmd = HistoryCommandFactory.createBatchCommand('Import NP SVG');
    const drawing = svgCanvas.getCurrentDrawing();
    const layerName = drawing.getCurrentLayerName();
    const layerModule = layerName ? getData(getLayerByName(layerName), 'module') : null;
    const elems = await importSvg(blob, {
      importType: 'layer',
      isFromNounProject: true,
      parentCmd: batchCmd,
      targetModule: layerModule,
    });

    if (!elems) return;

    for (const elem of elems) {
      await postImportElement(elem, batchCmd);
    }
    undoManager.addCommandToHistory(batchCmd);
  } finally {
    progressCaller.popById(progressId);
  }
};

const NPElement = ({ icon }: { icon: IIcon }) => {
  const { addToHistory, closeDrawer } = useContext(ElementPanelContext);
  const [isLoading, setIsLoading] = useState(true);
  const [hasError, setHasError] = useState(icon.hidden);

  if (hasError) return null;

  return (
    <div
      className={classNames(styles.icon, { [styles.loading]: isLoading })}
      id={icon.id}
      onClick={async () => {
        if (!isLoading) {
          addToHistory({ npIcon: icon, type: 'np' });
          await webNeedConnectionWrapper(async () => {
            await importNPSvg(icon.id);
            closeDrawer();
          });
        }
      }}
    >
      <img
        onError={() => {
          icon.hidden = true;
          setHasError(true);
        }}
        onLoad={() => setIsLoading(false)}
        src={icon.thumbnail_url}
      />
    </div>
  );
};

export default NPElement;
