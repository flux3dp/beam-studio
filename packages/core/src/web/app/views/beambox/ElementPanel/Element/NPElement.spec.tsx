import React from 'react';

import { fireEvent, render, waitFor } from '@testing-library/react';

const mockCloseDrawer = jest.fn();
const mockAddToHistory = jest.fn();

jest.mock('@core/app/contexts/ElementPanelContext', () => ({
  ElementPanelContext: React.createContext({ addToHistory: mockAddToHistory, closeDrawer: mockCloseDrawer }),
}));

const mockBatchCommand = { addSubCommand: jest.fn() };
const mockCreateBatchCommand = jest.fn().mockImplementation(() => mockBatchCommand);

jest.mock('@core/app/svgedit/history/HistoryCommandFactory', () => ({
  createBatchCommand: mockCreateBatchCommand,
}));

const mockAddCommandToHistory = jest.fn();

jest.mock('@core/app/svgedit/history/undoManager', () => ({
  addCommandToHistory: mockAddCommandToHistory,
}));

const mockElement = document.createElement('use');
const mockImportSvg = jest.fn().mockResolvedValue([mockElement]);

jest.mock('@core/app/svgedit/operations/import/importSvg', () => mockImportSvg);

const mockPostImportElement = jest.fn();

jest.mock('@core/app/svgedit/operations/import/postImportElement', () => mockPostImportElement);

const mockGetNPIconByID = jest.fn().mockResolvedValue('data:image/svg+xml;base64,1234');

jest.mock('@core/helpers/api/flux-id', () => ({
  getNPIconByID: mockGetNPIconByID,
}));

const mockGetData = jest.fn().mockReturnValue(5);

jest.mock('@core/helpers/layer/layer-config-helper', () => ({
  getData: mockGetData,
}));

const mockGetLayerByName = jest.fn().mockReturnValue('mock-layer-elem');

jest.mock('@core/helpers/layer/layer-helper', () => ({
  getLayerByName: mockGetLayerByName,
}));

const mockGetCurrentLayerName = jest.fn().mockReturnValue('mock-layer-name');

jest.mock('@core/helpers/svg-editor-helper', () => ({
  getSVGAsync: (callback) =>
    callback({
      Canvas: {
        getCurrentDrawing: () => ({ getCurrentLayerName: mockGetCurrentLayerName }),
      },
    }),
}));

jest.mock('@core/helpers/web-need-connection-helper', () => (callback) => callback());

const mockFetch = jest.fn().mockResolvedValue({ blob: () => Promise.resolve(new Blob()) });

global.fetch = mockFetch;

import NPElement from './NPElement';

describe('test NPElement', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    const { container } = render(<NPElement icon={{ id: '1234', thumbnail_url: 'img_thumbnail_url' }} />);

    expect(container).toMatchSnapshot();

    const icon = container.querySelector('.icon');
    const img = container.querySelector('img');

    fireEvent.click(icon);
    expect(mockAddToHistory).not.toHaveBeenCalled();

    fireEvent.load(img);
    expect(container).toMatchSnapshot();

    fireEvent.click(icon);
    waitFor(() => {
      expect(mockAddToHistory).toHaveBeenCalled();
      expect(mockAddToHistory).toHaveBeenCalledWith({
        npIcon: { id: '1234', thumbnail_url: 'img_thumbnail_url' },
        type: 'np',
      });
      expect(mockGetNPIconByID).toHaveBeenCalledWith('1234');
      expect(mockFetch).toHaveBeenCalledWith('data:image/svg+xml;base64,1234');
      expect(mockCreateBatchCommand).toHaveBeenCalledWith('Import NP SVG');
      expect(mockGetCurrentLayerName).toHaveBeenCalled();
      expect(mockGetLayerByName).toHaveBeenCalled();
      expect(mockGetData).toHaveBeenCalled();
      expect(mockImportSvg).toHaveBeenCalledWith(expect.any(Blob), {
        batchCmd: mockBatchCommand,
        importType: 'layer',
        isFromNounProject: true,
        targetModule: 5,
      });
      expect(mockPostImportElement).toHaveBeenCalledWith(mockElement, mockBatchCommand);
      expect(mockAddCommandToHistory).toHaveBeenCalledWith(mockBatchCommand);
      expect(mockCloseDrawer).toHaveBeenCalled();
    });
  });

  it('should render null when set hidden', async () => {
    const { container } = render(<NPElement icon={{ hidden: true, id: '1234', thumbnail_url: 'img_thumbnail_url' }} />);

    expect(container).toBeEmptyDOMElement();
  });

  it('should render null when img error', async () => {
    const { container } = render(<NPElement icon={{ id: '1234', thumbnail_url: 'img_thumbnail_url' }} />);
    const img = container.querySelector('img');

    fireEvent.error(img);
    expect(container).toBeEmptyDOMElement();
  });
});
