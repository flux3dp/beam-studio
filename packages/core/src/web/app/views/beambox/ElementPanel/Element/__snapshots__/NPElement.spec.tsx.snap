// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test NPElement should render correctly 1`] = `
<div>
  <div
    class="icon loading"
    id="1234"
  >
    <img
      src="img_thumbnail_url"
    />
  </div>
</div>
`;

exports[`test NPElement should render correctly 2`] = `
<div>
  <div
    class="icon"
    id="1234"
  >
    <img
      src="img_thumbnail_url"
    />
  </div>
</div>
`;
